### DISCUSSION TITLE
Alerts - how to add additional data to the 'contributingEvents'?

### DISCUSSION THREAD

**Discussion by kejdajar** [datalust/seq-tickets#1765] [Help] [answered]
Hello,

I am using the `Seq.App.EmailPlus` package to send alerts to a given e-mail address. For instance, whenever there is a burst of events such as warnings or errors, an aggregated e-mail is created. I have also enabled the `Include contributing events` options in the alert section.

As stated in the docs (https://docs.datalust.co/docs/alert-properties#contributing-events), the `ContributingEvents` contain:
- ID of the event
- Timestamp
- Message of the event

Is there any way to add additional data (such as `applicationName`) to the contributing events, so that it would be possible to display more information for each contributing event? (I have discovered that the frontend part of this could be achieved by altering the default HTML template for the `Seq.App.EmailPlus` app).

The issue is that the long list of contributing events does not provide an easy way how to tell which application the event is related to (without having to click on each link individually). Also, is there any way to group those "aggregated summary" e-mails by some other property, such as `applicationName`? (I guess this is not possible, because our property `applicationName` which works for normal alerts does not work (= has no value) in the aggregated alerts when being displayed in the title of the email)

The already mentioned `applicationName` property is in our case attached to the API key our app is using.
Thank you for your help.

Version: Seq 2023.1.8829

**Comment by nblumhardt** [comment, marked as answer]
Hi! Thanks for dropping us a line. Is your alert already grouping by `applicationName`? I.e. in:

![image](https://user-images.githubusercontent.com/342712/218611601-53296fde-9f85-4b9d-97f1-3dd255e6a33c.png)

This would change the meaning of the alert slightly (count of errors per application, instead of global count of errors) but it might make interpreting the resulting notifications easier.

There's no way to add additional properties to the contributing events at this point, unfortunately.

Let me know if this helps,
Nick

**Reply by kejdajar** [reply]
Thank you for the quick reply,

this is exactly what I needed. The last thing I need is to somehow put the `AppName` from the `Results` table into the subject of the e-mail.

I have found out that there should be `Source.Results` object available  (https://docs.datalust.co/docs/alert-properties#results). I can also confirm that this object is used in the `Seq.App.EmailPlus` to render the Results table.

I think that the title of the email can be customized to contain the `AppName`  property - maybe by using the `Notification message` section in the detail of the alert, or by setting the `Subject template` in the `Seq.App.EmailPlus` application - I have seen attempts like that in this thread: (https://github.com/datalust/seq-tickets/discussions/1649) 


Under the `Alert` section, there is a `Notification message` section - I tried to set the application name there (I was not sure about the curly braces so I tried more examples):

``My message - {Source.Results[0][0]} {{Source.Results[0][0]}} {{{Source.Results[0][0]}}}``

Within the `Seq.App.EmailPlus` I have also tried to access some data from the `Results` object within the `Subject template`:

``[DEV][{{$Level}}][{{AppName}}] - {{{$Message}}} {{Source.Results[0][0]}}``

The [0][0] is just an example, but I guess there should be some data in that cell.

My observations are:
- the `Source.Results` does not seem to work for me
- $Level, $Messages - always works
- {{AppName}} - works only for normal e-mails, but is empty in aggregated e-mails

The resulting e-mail looks like this (the variable `Source.Results` is not being replaced and the `AppName` property is empty):

![image](https://user-images.githubusercontent.com/22682824/218701299-97f10915-af04-47d7-ad84-ac780aa54fdd.png)

Thank you for your help.
(as a side note - if there is any way how to examine the actual data object the HTML template has recieved from the alert, let me know!)

**Reply by nblumhardt** [reply]
Hi! I think `Source.Results[0][0]` might instead need to be:

```
Source.Results.[0].[0]
```

I.e., there's a `.` before the indexer.

Let me know if this helps - fingers crossed :-)

Best regards,
Nick

**Reply by kejdajar** [reply]
I should have examined the Handlebars documentation more closely. One thing to remember - the ``Source.Results.[0].[0]`` works only in the `Seq.App.EmailPlus` Subject template, not in the Alerts/MyAlert/Notification message input.

Here is the final working *Subject Template*:
``[DEV] [{{$Level}}] [{{AppName}}{{Source.Results.[1].[1]}}] - {{#if Source.Results.[1].[1]}}Aggregated {{$Level}}s{{else}}{{{$Message}}}{{/if}}``

For aggregated warnings, the subject is:

- *[DEV] [Warning] [MyApplicationName] - Aggregated Warnings*

For individual warnings, the subject is:
- *[DEV] [Warning] [MyApplicationName] - My custom warning 1*

The same applies also to other log levels.
Thank you very much, I am closing the issue.

**Reply by nblumhardt** [reply]
Fantastic, thanks 👍
